drivers/flash/flash_npcx_fiu_nor.c

```diff
@@ -0,0 +1,609 @@
+	int ret = 0;
+	uint8_t opcode;
+
+	/* Out of the region of nor flash device? */
+	if ((addr < 0) || (((uint64_t)size + (uint64_t)addr) > config->flash_size)) {
+		LOG_ERR("Addr %ld, size %d are out of range", addr, size);
+		return -EINVAL;
+	}
+
+	/* address must be sector-aligned */
+	if (!SPI_NOR_IS_SECTOR_ALIGNED(addr)) {
+		LOG_ERR("Addr %ld is not sector-aligned", addr);
+		return -EINVAL;
+	}
+
+	/* size must be a multiple of sectors */
+	if ((size % config->min_erase_size) != 0) {
+		LOG_ERR("Size %d is not a multiple of sectors", size);
+		return -EINVAL;
+	}
+
+	/* Select erase opcode by size */
+	if (size == config->flash_size) {
+		opcode = SPI_NOR_CMD_CE;
+	} else if (config->min_erase_size == KB(4)) {
+		opcode = SPI_NOR_CMD_SE;
+	} else if (config->min_erase_size == KB(64)) {
+		opcode = SPI_NOR_CMD_BE;
+	} else {
+		return -EINVAL;
+	}
+
+	while (size > 0) {
+		flash_npcx_uma_cmd_only(dev, SPI_NOR_CMD_WREN);
+
+		if (opcode == SPI_NOR_CMD_CE) {
+			/* chip erase */
+			flash_npcx_uma_cmd_only(dev, opcode);
+			size -= config->flash_size;
+		} else {
+			flash_npcx_uma_cmd_by_addr(dev, opcode, addr);
+			addr += config->min_erase_size;
+			size -= config->min_erase_size;
+		}
+
+		ret = flash_npcx_nor_wait_until_ready(dev);
+		if (ret != 0) {
+			break;
+		}
+	}
+
+	return ret;
+}
+
+static int flash_npcx_nor_write(const struct device *dev, off_t addr,
+				  const void *data, size_t size)
+{
+	const struct flash_npcx_nor_config *config = dev->config;
+	uint8_t *tx_buf = (uint8_t *)data;
+	int ret = 0;
+
+	/* Out of the region of nor flash device? */
+	if ((addr < 0) || (((uint64_t)size + (uint64_t)addr) > config->flash_size)) {
+		return -EINVAL;
+	}
+
+	while (size > 0) {
+		size_t sz_write = size;
+
+		/* Don't write more than a page. */
+		if (sz_write >= SPI_NOR_PAGE_SIZE) {
+			sz_write = SPI_NOR_PAGE_SIZE;
+		}
+
+		/* Don't write across a page boundary */
+		if (((addr + sz_write - 1U) / SPI_NOR_PAGE_SIZE)
+							!= (addr / SPI_NOR_PAGE_SIZE)) {
+			sz_write = SPI_NOR_PAGE_SIZE - (addr % SPI_NOR_PAGE_SIZE);
+		}
+
+		/* Start to write */
+		flash_npcx_uma_cmd_only(dev, SPI_NOR_CMD_WREN);
+		ret = flash_npcx_uma_write_by_addr(dev, SPI_NOR_CMD_PP,
+						   tx_buf, sz_write, addr);
+		if (ret != 0) {
+			break;
+		}
+		size -= sz_write;
+		tx_buf += sz_write;
+		addr += sz_write;
+		/* Wait for writing completed */
+		ret = flash_npcx_nor_wait_until_ready(dev);
+		if (ret != 0) {
+			break;
+		}
+	}
+
+	return ret;
+}
+
+static const struct flash_parameters *
+flash_npcx_nor_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return &flash_npcx_parameters;
+};
+
+#ifdef CONFIG_FLASH_EX_OP_ENABLED
+static int flash_npcx_nor_ex_exec_uma(const struct device *dev, const uintptr_t in,
+					   void *out)
+{
+	const struct npcx_ex_ops_uma_in *op_in =
+		(struct npcx_ex_ops_uma_in *)in;
+	const struct npcx_ex_ops_uma_out *op_out =
+		(const struct npcx_ex_ops_uma_out *)out;
+	int flag = 0;
+#ifdef CONFIG_USERSPACE
```

Comment:
Non-blocking comment - can the USERSPACE checks get moved to `flash_npcx_nor_ex_op()` instead?  

Logically, the USERSPACE checks probably belong in  `flash_npcx_nor_ex_op()` because that routine is the part of driver API.  But I didn't fully check the logic whether all the `z_user_from_copy()` and `z_user_to_copy()` calls can be managed by wrapper routine.

Reply:
Sorry for the late reply. It sounds good. I will update it later.

Reply:
Done.